src/crab/wto.hpp

```diff
@@ -1,467 +1,180 @@
-
-  public:
-    using iterator = wto_component_list_t::iterator;
-    using const_iterator = wto_component_list_t::const_iterator;
-
-    vertex_descriptor_t head() { return this->_head; }
-
-    iterator begin() { return _wto_components.begin(); }
-
-    iterator end() { return _wto_components.end(); }
-
-    const_iterator begin() const { return _wto_components.begin(); }
-
-    const_iterator end() const { return _wto_components.end(); }
-
-    void increment_fixpo_visits() { _num_fixpo++; }
-
-    friend std::ostream& operator<<(std::ostream& o, const wto_cycle_t& cycle) {
-        o << "(" << cycle._head;
-        if (!cycle._wto_components.empty()) {
-            o << " ";
-            for (const wto_component_t& c : cycle) {
-                o << c << " ";
-            }
+    // Construct a Weak Topological Ordering from a control-flow graph using
+    // the algorithm of figure 4 in the paper, where this constructor matches
+    // what is shown there as the Partition function.
+    wto_t(const cfg_t& cfg) : _cfg(cfg) {
+        for (const label_t& label : cfg.labels()) {
+            _dfn.emplace(label, 0);
         }
-        o << ")";
-        if (cycle._num_fixpo > 0)
-            o << "^{" << cycle._num_fixpo << "}";
-        return o;
+        _num = 0;
+        visit(cfg.entry_label(), _components, {});
     }
 
-}; // class wto_cycle
-
-inline std::ostream& operator<<(std::ostream& o, const wto_vertex_t& vertex) {
-    return o << vertex._node;
-}
-
-inline std::ostream& operator<<(std::ostream& o, const wto_component_t& c) {
-    return std::visit(overloaded{
-        [&](const wto_cycle_t& x) -> std::ostream& { return o << x; },
-        [&](const wto_vertex_t& x) -> std::ostream& { return o << x; },
-    }, c);
-}
-
-class wto_t final {
-  private:
-    using wto_component_list_t = std::forward_list<wto_component_t>;
-    using dfn_t = bound_t;
-    using dfn_table_t = std::map<vertex_descriptor_t, dfn_t>;
-    using stack_t = std::vector<vertex_descriptor_t>;
-    using nesting_table_t = std::map<vertex_descriptor_t, wto_nesting_t>;
-
-    wto_component_list_t _wto_components;
-    dfn_table_t _dfn_table;
-    dfn_t _num{0};
-    stack_t _stack;
-    nesting_table_t _nesting_table;
-
-    class nesting_builder {
-      private:
-        wto_nesting_t _nesting;
-        nesting_table_t& _nesting_table;
-
-      public:
-        explicit nesting_builder(nesting_table_t& nesting_table) : _nesting_table(nesting_table) {}
-
-        void operator()(wto_cycle_t& cycle) {
-            vertex_descriptor_t head = cycle.head();
-            wto_nesting_t previous_nesting = this->_nesting;
-            this->_nesting_table.insert(std::make_pair(head, this->_nesting));
-            this->_nesting += head;
-            for (wto_component_t& c : cycle) {
-                std::visit(*this, c);
-            }
-            this->_nesting = previous_nesting;
+    [[nodiscard]] wto_partition_t& components() { return _components; }
```

Comment:
If possible, I think it's better to have a `begin()` and `end()` and to keep the order (and collection) of the components an implementation detail. It should make at least some of the loops simpler.

Reply:
Done